File: python/ft/strategy.py
import redis

import ft.order_sender as order_sender


class OrderResponse(object):
    def __init__(self, data):
        pass


class Tick(object):
    def __init__(self, data):
        pass


class Strategy(object):
    def __init__(self, strategy_id):
        self.strategy_id = strategy_id
        self.sender = order_sender.OrderSender(strategy_id=strategy_id)
        self.tick_redis = redis.StrictRedis(host='localhost', port=6379)
        self.tick_sub = self.tick_redis.pubsub()

    def on_init(self):
        pass

    def on_tick(self, tick):
        print('on_tick')

    def on_order_rsp(self, order):
        pass

    def on_exit(self):
        pass

    def subscribe(self, tickers):
        self.tick_sub.subscribe(tickers)

    def send_order(self, ticker, direction, offset, order_type, volume, price):
        self.sender.send_order(ticker, direction, offset,
                               order_type, volume, price)

    def run(self):
        self.on_init()

        if self.strategy_id:
            self.subscribe(Strategy)

        for reply in self.tick_sub.listen():
            if reply['channel'] == self.strategy_id:
                rsp = OrderResponse(reply['data'])
                self.on_order_rsp(rsp)
            else:
                tick = Tick(reply['data'])
                self.on_tick(tick)

File: src/cep/strategy/strategy.cpp
// Copyright [2020] <Copyright Kevin, kevin.lau.gd@gmail.com>

#include "strategy/strategy.h"

namespace ft {

void Strategy::run() {
  on_init();
  puller_.subscribe_order_rsp(strategy_id_);

  for (;;) {
    auto reply = puller_.pull();
    if (reply) {
      if (strcmp(reply->element[1]->str, strategy_id_) == 0) {
        auto rsp =
            reinterpret_cast<const OrderResponse*>(reply->element[2]->str);
        on_order_rsp(*rsp);
      } else {
        auto tick = reinterpret_cast<TickData*>(reply->element[2]->str);
        on_tick(*tick);
      }
    }
  }
}

void Strategy::subscribe(const std::vector<std::string>& sub_list) {
  puller_.subscribe_md(sub_list);
}

}  // namespace ft
